src/webgpu/shader/execution/expression/binary/i32_arithmetic.spec.ts

```diff
@@ -0,0 +1,180 @@
+export const description = `
+Execution Tests for the i32 arithmetic binary expression operations
+`;
+
+import { makeTestGroup } from '../../../../../common/framework/test_group.js';
+import { GPUTest } from '../../../../gpu_test.js';
+import { kValue } from '../../../../util/constants.js';
+import { i32, TypeI32 } from '../../../../util/conversion.js';
+import { cartesianProduct, fullI32Range } from '../../../../util/math.js';
+import { makeCaseCache } from '../case_cache.js';
+import { allInputSources, Case, run } from '../expression.js';
+
+import { binary } from './binary.js';
+
+/**
+ * A function that performs a binary operation on x and y, and returns the expected
+ * result, or undefined if the operation is invalid for the given inputs.
+ */
+interface BinaryToI32Op {
+  (x: number, y: number): number | undefined;
+}
+
+/**
+ * @returns an array of Cases for operations over a range of inputs
+ * @param param0s array of inputs to try for the first param
+ * @param param1s array of inputs to try for the second param
+ * @param op callback called on each pair of inputs to produce each case
+ */
+function generateBinaryToI32Cases(params0s: number[], params1s: number[], op: BinaryToI32Op) {
+  return cartesianProduct(params0s, params1s).reduce((cases, e) => {
+    const expected = op(e[0], e[1]);
+    if (expected !== undefined) {
+      cases.push({ input: [i32(e[0]), i32(e[1])], expected: i32(expected) });
+    }
+    return cases;
+  }, new Array<Case>());
+}
```

Comment:
I could move these to expression.ts, where the analogous f32 interface and function live, but since it's only being used here, I thought I'd keep it here for now. Let me know if you want me to move it.

Reply:
I think I would prefer it over in `expression.ts`.

Reply:
Done.